plugin-process-to-cgroup-bridge/src/transform.rs

```diff
@@ -0,0 +1,199 @@
+        let procfs_cgroup_base_path = get_proc_path();
+
+        let procfs_cgroup_filepath = procfs_cgroup_base_path.join(pid.to_string()).join("cgroup");
+
+        if !Path::new(&procfs_cgroup_filepath).exists() {
+            return Ok(None);
+        }
+
+        let contents = fs::read_to_string(&procfs_cgroup_filepath)
+            .with_context(|| format!("Failed to read {:?}", procfs_cgroup_filepath))?;
+
+        // a typical procfs cgroup file will contain only one line
+        // eg: 0::/system.slice/docker-7c7fc86f5f2a609c41c6edd65bd1b64135124a687fa6516f6b177b040d6e3b68.scope
+        for line in contents.lines() {
+            let parts: Vec<&str> = line.split(':').collect();
+            if parts.len() >= 3 {
+                let cgroup_path = parts[2];
+                if !cgroup_path.is_empty() {
+                    return Ok(Some(cgroup_path.to_string()));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    fn extract_process_id_from_measurement(measurement: &MeasurementPoint) -> anyhow::Result<u32> {
+        match measurement.consumer {
+            ResourceConsumer::Process { pid } => Ok(pid),
+            _ => Err(anyhow!(
+                "expected a process resource consumer, got something else: {:?}",
+                measurement.consumer
+            )),
+        }
+    }
+
+    /// Aggregates measurements with the same metric, consumer and timestamp by calculating their mean value.
+    /// Groups are identified by (metric_id, consumer, timestamp) and all measurements in each
+    /// group are averaged together.
+    fn aggregate_cgroups_measurements(measurements: &mut MeasurementBuffer) -> MeasurementBuffer {
+        let mut grouped: HashMap<(RawMetricId, ResourceConsumer, u64), Vec<&MeasurementPoint>> = HashMap::new();
+
+        // Group by (metric_id, consumer, timestamp)
+        for point in measurements.iter() {
+            let ts = point.timestamp.duration_since(UNIX_EPOCH.into()).unwrap().as_secs();
+            let key = (point.metric, point.consumer.clone(), ts);
+            grouped.entry(key).or_default().push(point);
+        }
+
+        let mut aggregates = MeasurementBuffer::new();
+
+        // Calculate mean value for every group and create aggregated point
+        for ((_metric, _consumer, _timestamp), group) in grouped {
+            let (sum, count) = group
+                .iter()
+                .filter_map(|p| extract_numeric_value(p))
+                .fold((0.0, 0), |(sum, count), value| (sum + value, count + 1));
+
+            let mean = if count > 0 { sum / count as f64 } else { 0.0 };
+
+            let mut new_point = group[0].clone(); // cannot panic since group cannot be empty
+            new_point.value = WrappedMeasurementValue::F64(mean);
+            aggregates.push(new_point);
+        }
+        aggregates
+    }
+}
+
+fn extract_numeric_value(measurement: &MeasurementPoint) -> Option<f64> {
```

Comment:
The other "utilities" functions could be outside of the impl too, this way you would not need `Self::` to call them.